rust/actyx/ax-aql/src/language/aql.pest

```diff
@@ -7,23 +7,68 @@ main_timestamp = @{ SOI ~ isodate ~ EOI }
 
 // pragmas
 pragma = ${ "PRAGMA" ~ WHITE_SPACE+ ~ feature_word ~ ( !NEWLINE ~ WHITE_SPACE )* ~ ( pragma_multi | pragma_single ) }
-pragma_multi = _{ NEWLINE ~ pragma_multi_text ~ NEWLINE ~ "ENDPRAGMA" ~ NEWLINE }
-pragma_multi_text = ${ ( !( NEWLINE ~ "ENDPRAGMA" ~ NEWLINE ) ~ ANY )* }
+pragma_multi = _{ NEWLINE ~ pragma_multi_text ~ NEWLINE ~ pragma_end ~ NEWLINE }
+pragma_multi_text = ${ ( !( NEWLINE ~ pragma_end ~ NEWLINE ) ~ ANY )* }
+pragma_end = { "ENDPRAGMA" }
 pragma_single = _{ ":=" ~ ( !NEWLINE ~ WHITE_SPACE )* ~ pragma_single_text ~ NEWLINE }
 pragma_single_text = ${ ( !NEWLINE ~ ANY )* }
 
+features = { "FEATURES(" ~ feature_word* ~ ")" }
+feature_word = @{ ( ASCII_ALPHANUMERIC | "ø" )+ }
+
+// types
+type = { type_atom ~ type_post* ~ ( type_infix ~ type_atom ~ type_post* )* }
+type_null = { "NULL" }
+type_bool = { "BOOLEAN" }
+type_number = { "NUMBER" }
+type_timestamp = { "TIMESTAMP" }
+type_string = { "STRING" }
+type_universal = { "UNIVERSAL" }
+type_tuple = { brackl ~ type ~ ( comma ~ type )* ~ brackr }
+type_record = { curlyl ~ ( ( natural | ident | nonempty_string ) ~ colon ~ type ~ comma? )+ ~ curlyr }
+type_paren = { parenl ~ type ~ parenr }
+type_atom = _{ type_null | type_bool | bool | type_number | natural | type_timestamp | type_string | string
+  | type_tuple | type_record | type_paren | type_universal }
+type_array = { "[]" }
+type_dict = { "{}" }
+type_post = _{ type_array | type_dict }
+type_union = { "|" }
+type_intersection = { "&" }
+type_infix = _{ type_union | type_intersection }
+
+// events
+event = { "EVENT" ~ ident ~ event_type ~ tagged? }
+event_type = { type_record } // needed for parser code structure
+tagged = { "TAGGED" ~ tag ~ ( comma ~ tag )* }
+
+// workflows
+workflow = { "WORKFLOW" ~ ident ~ parenl ~ wf_arg ~ ( comma ~ wf_arg )* ~ parenr ~ wf_scope }
+wf_arg = { ( role | unique ) ~ ident }
+wf_scope = { curlyl ~ wf_step+ ~ curlyr }
+wf_step = _{ wf_event | wf_retry | wf_timeout | wf_parallel | wf_call | wf_compensate | wf_choice }
+wf_event = { ( wf_state ~ colon )? ~ ( fail | return )? ~ ident ~ at ~ ident ~ wf_binder* }
+wf_state = { ident }
+wf_binder = { curlyl ~ ident ~ colon ~ ident ~ arrowl ~ simple_expr ~ curlyr }
+wf_retry = { retry ~ wf_scope }
+wf_timeout = { timeout ~ duration ~ wf_scope ~ wf_event }
+wf_parallel = { parallel ~ natural? ~ wf_cases }
+wf_call = { match ~ ident ~ parenl ~ ident ~ ( comma ~ ident )* ~ parenr ~ curlyl ~ wf_match* ~ curlyr }
+wf_match = { case ~ ( ident | star ) ~ dblarrowr ~ wf_step+ }
```

Comment:
nitpick for UX reasons: I imagine `wf_step+` here will look ambiguous. it indicates two syntaxes of the atomicity of a block/function body: 1.) being encapsulated inside `{}` and 2.) otherwise.

Switching between the two may cause unintentional cognitive load.

Shouldn't it be a `(wf_step | wf_scope)` instead, or for some simplicity `wf_scope` only?

Reply:
I played with it a bit and since each case is terminated either by `CASE` or `}` I didn’t find it ambiguous.

But your questions reminds me: `wf_scope` isn’t actually about scoping, but I only figured that out after I’d started writing the grammar. Should probably be called `wf_block`.